### ISSUE TITLE
ANSI coded output causes Junit formatter to create invalid XML

### ISSUE THREAD

**Issue by bioball** [bats-core/bats-core#311]
If outputting text with ANSI controls, the Junit formatter will create invalid XML. For instance:

```bats
@test "my-test" {
  printf "\e[1mhello\e[0m"
  return 1
}
```

Produces this output (with properties redacted)

```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuite name="my_test.bats" tests="1" failures="1" errors="0" skipped="0" time="0" timestamp="2020-05-12T15:01:25">
    <testcase classname="my_test.bats" name="my-test" time="0">
        <failure type="failure"><![CDATA[
(in test file test/my_test.bats, line 106)
  `printf "\e[1mhello\e[0m"' failed
[1mhello[0m]]></failure>
    </testcase>
  <system-out></system-out>
  <system-err></system-err>
</testsuite>
```

The CDATA section includes the control character literal `` (Unicode: 0x1b), which are invalid XML characters. This is causing some XML parsers to fail parsing the resulting file.

I think maybe there's two ways to handle this:
1. Encode the escape character (as `\e` or `\033` or something)
2. Strip all control characters from the captured output

**Comment by sublimino** [comment]
Thanks for the report @bioball, and for kicking the tyres on this one : ]

cc @martin-schulze-vireso WDYT? Option 1 may add complexity when escaping for different targets, or might be as simple as a one-time escape before writing CDATA.

If the second option, I've been using a sed control-character replacer for years that may be suitable:

```
colour-strip () { 
    sed "s,\x1B\[[0-9;]*[a-zA-Z],,g"
}
```

**Comment by martin-schulze-vireso** [comment]
>  Option 1 may add complexity when escaping for different targets

What do you mean by different targets? My understanding is that the XML definition clearly states which symbols are allowed, so there really should be only "one rule" for that.

I think stripping them is good enough until someone complains about missing them.

However, this issue raises the question: What if someone does `printf "]]>"`? IMHO this hints at a more general requirement to properly sanitize/escape the output for the CDATA area.

**Comment by sublimino** [comment]
> What do you mean by different targets? 
Viewed on different operating systems and terminal types. Nothing specific, just general caution!

> I think stripping them is good enough until someone complains about missing them.
Agreed

> However, this issue raises the question: What if someone does printf "]]>"? IMHO this hints at a more general requirement to properly sanitize/escape the output for the CDATA area.

That's a good point, I don't know enough about XML to help here.

**Comment by martin-schulze-vireso** [comment]
> Viewed on different operating systems and terminal types. Nothing specific, just general caution!

I'd say that this should be done on a case by case basis. We cannot possible anticipate all kinds of "broken" behavior. The best we could do would be to restrict the output to printable ASCII or some other very small common denominator. Yet, even that would not prevent the CDATA problem.

> That's a good point, I don't know enough about XML to help here.

Me neither. What I gather from [this](https://stackoverflow.com/q/223652) is: As far as escaping is concerned we only need to deal with `]]>` and it can be replaced by `]]]]><![CDATA[>`, conforming parsers should merge the two adjacent occurences into one string.

**Comment by martin-schulze-vireso** [comment]
@sublimino Regarding your sed script: Can we assume to always have sed available?

Furthermore, do we want to remove the full color code or only the 0x1b char? If the latter is the case: Could a simple `tr -d "\033"` achieve that on a wider range of systems?

**Comment by sublimino** [comment]
@martin-schulze-vireso I've noticed it's already in `bats` under a test helper with the name `filter_control_sequences`, so `sed`'s already a test dependency. 

However (!) a non-sed solution would definitely be preferable, you're right. `coreutils` is better : ]

As for full colour code vs `0x1b` — I don't feel strongly. If we don't sanitise the whole string we may risk platform specific encoding/output bugs as I previously noted, but that's just supposition and I'm happy to just remove that char if that's the preferred approach.

**Comment by martin-schulze-vireso** [comment]
I don't have very strong opinions on that either. Retaining the color codes makes the output a bit less readable but somebody might be interested in them?

**Comment by bioball** [comment]
Thoughts:

1. Totally removing all controls makes sense, and probably helps with readability if the tools that consume these junit reports don't do any processing for them
2. If including ANSI controls, it seems like the proper way to do this is to use `&#27;` (see https://stackoverflow.com/questions/4134438/saving-an-escape-character-0x1b-in-an-xml-file)

It probably makes to strip ANSI by default, and maybe introduce a flag that encodes the control character if people ask for it.

**Comment by sublimino** [comment]
This has been moved to the `1.3.0` roadmap.

**Comment by martin-schulze-vireso** [comment]
I think I'd opt for totally removing the control character then, it is easier on our part with `tr -d`.

I think this could be lumpled together with either #341 and #342 because it touches on the same code area.

**Comment by jeff-cook** [comment]
From `pretty output with junit formatter` #342
> Unfortunately, no. They are both formatters and there can only be one formatter. However, during implementation the discussion already came up, whether junit output should be considered a formatter at all, because its real output is to files instead of stdout.

I could see not having junit being a formatter and have it a new `--report` output. That would be the best of both worlds to me.

**Comment by martin-schulze-vireso** [comment]
This has been fixed with #345.
